pylops_mpi/basicoperators/BlockDiag.py

```diff
@@ -0,0 +1,139 @@
+            Operator shape
+        explicit : :obj:`bool`
+            Operator contains a matrix that can be solved explicitly (``True``) or not (``False``)
+
+        Notes
+        -----
+        An MPI Block Diagonal operator is composed of N linear operators, represented by **L**.
+        Each operator performs forward mode operations using its corresponding data, denoted as **m**.
+        The data is a DistributedArray located at each rank.
+        The forward mode of each operator is then collected from all ranks as a DistributedArray, referred to as **d**.
+
+        .. math::
+          \begin{bmatrix}
+            \mathbf{d}_1 \\
+            \mathbf{d}_2 \\
+            \vdots \\
+            \mathbf{d}_n
+          \end{bmatrix} =
+          \begin{bmatrix}
+            \mathbf{L}_1 & \mathbf{0} & \ldots & \mathbf{0} \\
+            \mathbf{0} & \mathbf{L}_2 & \ldots & \mathbf{0} \\
+            \vdots & \vdots & \ddots & \vdots \\
+            \mathbf{0} & \mathbf{0} & \ldots & \mathbf{L}_n
+          \end{bmatrix}
+          \begin{bmatrix}
+            \mathbf{m}_1 \\
+            \mathbf{m}_2 \\
+            \vdots \\
+            \mathbf{m}_n
+          \end{bmatrix}
+
+        Likewise, for the adjoint mode, each operator executes operations in the adjoint mode,
+        and the results are gathered from each rank to form a DistributedArray represented by **d**.
+
+        .. math::
+          \begin{bmatrix}
+            \mathbf{d}_1 \\
+            \mathbf{d}_2 \\
+            \vdots \\
+            \mathbf{d}_n
+          \end{bmatrix} =
+          \begin{bmatrix}
+            \mathbf{L}_1^H & \mathbf{0} & \ldots & \mathbf{0} \\
+            \mathbf{0} & \mathbf{L}_2^H & \ldots & \mathbf{0} \\
+            \vdots & \vdots & \ddots & \vdots \\
+            \mathbf{0} & \mathbf{0} & \ldots & \mathbf{L}_n^H
+          \end{bmatrix}
+          \begin{bmatrix}
+            \mathbf{m}_1 \\
+            \mathbf{m}_2 \\
+            \vdots \\
+            \mathbf{m}_n
+          \end{bmatrix}
+
+    """
+
+    def __init__(self, ops: Sequence[LinearOperator],
+                 base_comm: MPI.Comm = MPI.COMM_WORLD,
+                 dtype: Optional[DTypeLike] = None):
+        # Required for MPI
+        self.base_comm = base_comm
+        self.rank = base_comm.Get_rank()
+        self.size = base_comm.Get_size()
+        self.ops = ops
+        mops = np.zeros(len(self.ops), dtype=np.int64)
+        nops = np.zeros(len(self.ops), dtype=np.int64)
+        for iop, oper in enumerate(self.ops):
+            nops[iop] = oper.shape[0]
+            mops[iop] = oper.shape[1]
+        self.mops = mops.sum()
+        self.nops = nops.sum()
+        self.nnops = np.insert(np.cumsum(nops), 0, 0)
+        self.mmops = np.insert(np.cumsum(mops), 0, 0)
+        # Shape of the operator is equal to the sum of nops and mops at all ranks
+        self.shape = (self.base_comm.allreduce(self.nops), self.base_comm.allreduce(self.mops))
```

Comment:
you should not do self.shape, and then pass it to super... make just a temporary shape and then pass it. Same for dtype (basically the way you did clinear is how you should do all)

Reply:
Done....